### ISSUE TITLE
Support for Aqara Door Sensor

### ISSUE THREAD

**Issue by pierrecle** [aholstenson/miio#101]
Subdevices `53` are also magnets (Aqara Door Sensor).
But it seams that the sensor doesn't get its contact value at startup. My door is **closed** but:
```
child.isOpen().then((o) => console.log(`isOpen:${o}`)); // => isOpen:true
child.on('opened', () => console.log(`${child.internalId} opened`)); // => ok here
child.on('closed', () => console.log(`${child.internalId} closed`)); // => ok here
```
Oh and of course, to test that, I added the following line in `subdevices.js`:
```
53: require('./magnet')
```
Any idea of why it doesn't get its value at start?

**Comment by aholstenson** [comment]
Thanks for opening an issue! Some values are only available from the gateway via the developer API which `miio` has only used for getting reports. I've spent some time making sure that it's now also used to read initial values. The current master should now properly report the initial open/close state.

I'll keep this issue open until this is released to NPM, which hopefully will be today or tomorrow.

**Comment by pierrecle** [comment]
Thanks! I tried on the master branch the following simple script:
```js
const miio = require('./lib');

// Create a new device over the given address
miio.devices().on('available', reg => {
	if(!reg.token) {
	  return;
    }
    
	const device = reg.device;
	if(device && device.matches('type:miio:gateway')) {
		const children = device.children();
		for(const child of children) {
			if(child.matches('type:miio:subdevice') && child.matches('cap:contact')) {
				console.log('magnet');
				child.on('opened', () => console.log("opened"));
				child.on('closed', () => console.log("opened"));
				child.isOpen().then((o) => console.log('o:' + o));
			}
		}
	}
  });
```
But I have the following error:
```(node:83644) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): undefined```

I have a gateway, an aqara door sensor and an aqara weather connected to it.

**Comment by aholstenson** [comment]
Looks like the read-command over the developer API didn't complete within 5 seconds. I've pushed a few more changes that might help with debugging this.

Could you try running with the `DEBUG` flag set to something like `thing:miio\*` to see if you are receiving responses to the initial read calls? For most shells something like `DEBUG=thing:miio\* node scriptHere.js` should allow you to see lots of debug info.

**Comment by pierrecle** [comment]
Thanks, here is the output:
```
  thing:miio:77249638 <- Handshake reply: <Buffer dd b7 81 7d 84 5d 8f a0 e1 fe d6 93 ea 2a 57 4a> +0ms
  thing:miio:77249638 -> (5) {"id":1,"method":"miIO.info","params":[]} +7ms
  thing:miio:77249638 <- Message: `{"result":{"life":6800,"cfg_time":0,"token":"ddb7817d845d8fa0e1fed693ea2a574a","mac":"78:11:DC:A4:7C:1D","fw_ver":"1.4.1_151","hw_ver":"MW300","model":"lumi.gateway.v3","mcu_fw_ver":"0143","wifi_fw_ver":"SD878x-14.76.36.p84-702.1.0-WM","ap":{"rssi":-29,"ssid":"XXXXX","bssid":"XX:XX:XX:XX:XX:XX"},"netif":{"localIp":"192.168.2.249","mask":"255.255.255.0","gw":"192.168.2.1","gw_mac":"00:0D:B4:14:49:AC"},"mmfree":176448,"ot":"otu","otu_stat":[314,319,261,0,247,419],"ott_stat":[0, 0, 0, 0]},"id":1}` +9ms
  thing:miio:77249638 -> (5) {"id":2,"method":"get_prop","params":["illumination","rgb"]} +20ms
  thing:miio:77249638 <- Message: `{"result":[1292,0],"id":2}` +4ms
  thing:miio:77249638 Property illuminance changed from undefined to 1292 +0ms
  thing:miio:77249638 Property rgb changed from undefined to { red: 0, green: 0, blue: 0 } +2ms
  thing:miio:77249638 Property brightness changed from undefined to 0 +1ms
  thing:miio:77249638 -> (5) {"id":3,"method":"get_lumi_dpf_aes_key","params":[]} +4ms
  thing:miio:77249638 <- Message: `{"result":["74FED41A48164F24"],"id":3}` +4ms
  thing:miio:77249638 -> (5) {"id":4,"method":"get_device_prop","params":["lumi.0","device_list"]} +1ms
  thing:miio:77249638 <- Message: `{"result":["lumi.158d0001a4b965",19,1,0,3,"lumi.158d0001b1458c",53,1,0,3],"id":4}` +4ms
  thing:miio:77249638 DEV -> {"cmd":"read","sid":"158d0001a4b965"} +16ms
  thing:miio:77249638 DEV -> {"cmd":"read","sid":"158d0001b1458c"} +0ms
  thing:miio:77249638 <- Handshake reply: <Buffer dd b7 81 7d 84 5d 8f a0 e1 fe d6 93 ea 2a 57 4a> +450ms
  thing:miio:77249638 miio:158d0001a4b965 got error: Error: Timeout while waiting for sub-device properties
    at Timeout.setTimeout (/Users/Pierre/Development/miio/lib/devices/gateway/subdevice.js:173:12)
    at ontimeout (timers.js:386:14)
    at tryOnTimeout (timers.js:250:5)
    at Timer.listOnTimeout (timers.js:214:5) +5s
  thing:miio:77249638 miio:158d0001b1458c got error: Error: Timeout while waiting for sub-device properties
    at Timeout.setTimeout (/Users/Pierre/Development/miio/lib/devices/gateway/subdevice.js:173:12)
    at ontimeout (timers.js:386:14)
    at tryOnTimeout (timers.js:250:5)
    at Timer.listOnTimeout (timers.js:214:5) +3ms
  thing:miio:77249638 DEV <- {"cmd":"heartbeat","model":"gateway","sid":"7811dca47c1d","short_id":"0","token":"4PmObnlN0jZESAR9","data":"{\"ip\":\"192.168.2.249\"}"} +3s
```

**Comment by pierrecle** [comment]
I tested with 60s timeout, same fail. I'm not sure that `magnet` send its value initially, a manual `read` command might be required. But if I read well the code, it's already done, right?

**Comment by aholstenson** [comment]
Yes, the `read` command is sent via the developer API but it doesn't look like you are receiving a reply. But the heartbeats and reports seem to be coming in correctly.

In my case I get a reply a few milliseconds after:
```
DEV <- {"cmd":"read_ack","model":"magnet","sid":"158d0001592c79","short_id":54235,"data":"{\"voltage\":3025,\"status\":\"open\"}"}
```

Right now I'm thinking that the port of the gateway might have been the issue, I've just pushed some code to use the developer API to detect it. Give it a try and see if it helps, otherwise I'm thinking that I'll add a fallback to use the old behavior.

**Comment by pierrecle** [comment]
Same after update, but I think that my wifi network is particular here:
```
INFO  Discovering devices. Press Ctrl+C to stop.

  thing:miio:77249638 <- Handshake reply: <Buffer dd b7 81 7d 84 5d 8f a0 e1 fe d6 93 ea 2a 57 4a> +0ms
  thing:miio:77249638 -> (5) {"id":1,"method":"miIO.info","params":[]} +6ms
  thing:miio:77249638 <- Message: `{"result":{"life":9940,"cfg_time":0,"token":"ddb7817d845d8fa0e1fed693ea2a574a","mac":"78:11:DC:A4:7C:1D","fw_ver":"1.4.1_151","hw_ver":"MW300","model":"lumi.gateway.v3","mcu_fw_ver":"0143","wifi_fw_ver":"SD878x-14.76.36.p84-702.1.0-WM","ap":{"rssi":-31,"ssid":"XXXXXXXXXX","bssid":"XX:XX:XX:XX:XX:XX"},"netif":{"localIp":"192.168.2.249","mask":"255.255.255.0","gw":"192.168.2.1","gw_mac":"00:0D:B4:14:49:AC"},"mmfree":176448,"ot":"otu","otu_stat":[284,299,282,0,268,419],"ott_stat":[0, 0, 0, 0]},"id":1}` +9ms
  thing:miio:77249638 -> (5) {"id":2,"method":"get_prop","params":["illumination","rgb"]} +27ms
  thing:miio:77249638 <- Message: `{"result":[1292,0],"id":2}` +5ms
  thing:miio:77249638 Property illuminance changed from undefined to 1292 +0ms
  thing:miio:77249638 Property rgb changed from undefined to { red: 0, green: 0, blue: 0 } +1ms
  thing:miio:77249638 Property brightness changed from undefined to 0 +2ms
  thing:miio:77249638 -> (5) {"id":3,"method":"get_lumi_dpf_aes_key","params":[]} +5ms
  thing:miio:77249638 <- Message: `{"result":["74FED41A48164F24"],"id":3}` +5ms
  thing:miio:77249638 DEV BROADCAST -> {"cmd":"whois"} +8ms
  thing:miio:77249638 <- Handshake reply: <Buffer dd b7 81 7d 84 5d 8f a0 e1 fe d6 93 ea 2a 57 4a> +453ms
  thing:miio:77249638 -> (5) {"id":4,"method":"get_device_prop","params":["lumi.0","device_list"]} +556ms
  thing:miio:77249638 <- Message: `{"result":["lumi.158d0001a4b965",19,1,0,3,"lumi.158d0001b1458c",53,1,0,3],"id":4}` +4ms
  thing:miio:77249638 DEV -> {"cmd":"read","sid":"158d0001a4b965"} +1s
  thing:miio:77249638 DEV -> {"cmd":"read","sid":"158d0001b1458c"} +0ms
  thing:miio:77249638 miio:158d0001a4b965 got error: Error: Timeout while waiting for sub-device properties
    at Timeout.setTimeout (/Users/Pierre/Development/miio/lib/devices/gateway/subdevice.js:173:12)
    at ontimeout (timers.js:386:14)
    at tryOnTimeout (timers.js:250:5)
    at Timer.listOnTimeout (timers.js:214:5) +5s
  thing:miio:77249638 miio:158d0001b1458c got error: Error: Timeout while waiting for sub-device properties
    at Timeout.setTimeout (/Users/Pierre/Development/miio/lib/devices/gateway/subdevice.js:173:12)
    at ontimeout (timers.js:386:14)
    at tryOnTimeout (timers.js:250:5)
    at Timer.listOnTimeout (timers.js:214:5) +1ms
  thing:miio:77249638 DEV <- {"cmd":"heartbeat","model":"gateway","sid":"7811dca47c1d","short_id":"0","token":"tjeqirwh4hGYSc8O","data":"{\"ip\":\"192.168.2.249\"}"} +2s
  thing:miio:77249638 DEV <- {"cmd":"heartbeat","model":"gateway","sid":"7811dca47c1d","short_id":"0","token":"EaceNdmkfR4toFVZ","data":"{\"ip\":\"192.168.2.249\"}"} +10s
  thing:miio:77249638 DEV <- {"cmd":"heartbeat","model":"gateway","sid":"7811dca47c1d","short_id":"0","token":"cEQJGR5mIotR9TR1","data":"{\"ip\":\"192.168.2.249\"}"} +10s
  thing:miio:77249638 -> (5) {"id":5,"method":"get_prop","params":["illumination","rgb"]} +29s
  thing:miio:77249638 <- Message: `{"result":[1292,0],"id":5}` +4ms
  thing:miio:77249638 DEV <- {"cmd":"heartbeat","model":"gateway","sid":"7811dca47c1d","short_id":"0","token":"Rln2BPMxayTrVEc0","data":"{\"ip\":\"192.168.2.249\"}"} +10s
  thing:miio:77249638 DEV <- {"cmd":"heartbeat","model":"gateway","sid":"7811dca47c1d","short_id":"0","token":"KeNEtqJiVuGFq6Qk","data":"{\"ip\":\"192.168.2.249\"}"} +10s
  thing:miio:77249638 DEV <- {"cmd":"heartbeat","model":"gateway","sid":"7811dca47c1d","short_id":"0","token":"p7eFgnb2zsyP39Z7","data":"{\"ip\":\"192.168.2.249\"}"} +10s
  thing:miio:77249638 -> (5) {"id":6,"method":"get_prop","params":["illumination","rgb"]} +30s
  thing:miio:77249638 <- Message: `{"result":[1292,0],"id":6}` +5ms
```
I'll test this evening at home. Thanks!

**Comment by pierrecle** [comment]
I tried with the master branch, at home with a good network quality, and same problem as above, with my 2 gateways.

**Comment by aholstenson** [comment]
The issue in your repo where you linked back here got me thinking. Are you running any other code that uses the Developer API while testing miio? In that case both libraries will be trying to bind to the same port which might be causing issues. This library sets `reuseAddr` so it should work fine if other libraries do the same.

**Comment by pierrecle** [comment]
On my laptop for development nope, only 1 code. For now, I'm only running `miio` until everything works for me, then I'll continue my integration 😃 
It's very strange because yesterday everything was fine. I tried with your latest commits, now I don't have any error but no output when using discover. Maybe the change on the ports? I'll try that.

**Comment by aholstenson** [comment]
I just pushed some code that uses the old way to load some initial properties if the developer API doesn't respond in a second or two. That should make things like temperature/humidity sensor work as intended even if you run into this issue. The door sensor will still not report the proper initial status if the fallback need to run.

**Comment by pierrecle** [comment]
You're right, everything run as expected, except the initial status of the aqara door as you mentioned. I also search a way do fix that.

**Comment by pierrecle** [comment]
I'll create another issue for the default magnet state, to close this one once the new version of the NPM package is published.

**Comment by aholstenson** [comment]
Just published 0.15.3. Closing this issue.
